File: fast-shop-API/src/main/kotlin/com/fastshop/domain/model/Product.kt
package com.fastshop.domain.model

import com.fastshop.domain.exception.BusinessException
import java.math.BigDecimal
import javax.persistence.*

@Entity
data class Product(
    @Id
    @GeneratedValue(strategy=GenerationType.IDENTITY)
    var id:Long?=null,
    val name:String="",
    val description:String="",
    @OneToOne
    var category:Category?=null,
    var availableUnities:Long?=0,
    @Enumerated(EnumType.STRING)
    val measurementUnit: MeasurementUnity?=null,

    val price:BigDecimal= BigDecimal.ZERO,
    val purchisePrice:BigDecimal= BigDecimal.ZERO,
    ){

    fun increaseQuantity(qt:Int){
        if(qt<1) throw BusinessException("Invalid Stock quntity")
        this.availableUnities = this.availableUnities?.plus(qt)
    }

}

File: fast-shop_UI/src/app/core/models/Product.ts
import { Link, Page } from "../http/Responses"
import { Category } from "./Category"

export interface Product{
  id:Number
  name:string,
  description:string
  measurement_unit:string
  available_unities: number
  price:Number
  purchise_price:Number
  category:Category
  links:Link[]
}

export interface ProductCollection{
  content:Product[],
  links:Link[]
  page?:Page
}

export interface ProducInput{
 name:string,
 description:String
 measurement_unit:string
 price:Number
 reseller_price:Number
 purchise_price:Number
 category_id:Number
 self_url?:string
}
